test(protocol-designer): e2e batch edit automation cases

reachojhaneha opened the pull request:
# Overview

Writing the e2e tests for batch

# Changelog

- Automation cases for batch edit


# Risk assessment
low

Comment
work in progress, this PR needs feature flag to be removed. Blocked by #7446

Commit:
batch edit e2e

Commit:
pulling latest from edge

Commit:
test(protocol-designer): e2e batch edit automation cases

closes #7384

#7384

Commit:
removing the errors

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
adding the multiselect

IanLondon committed:
feat(components, protocol-designer): provide more e2e selectors

Addresses #7384

- add id attr to TitleBar, use in PD's TitleBar #TitleBar_main
- pass through name attr for CheckboxRowField down to the checkbox input
- add id #Text_noSharedSettings to the "no shared settings" text for
  Batch Edit Mode (when you have multiple step types selected!)

IanLondon committed:
Merge remote-tracking branch 'origin/edge' into pd_add-e2e-selectors-7384

IanLondon committed:
pass in WellOrder field names

IanLondon committed:
add name attrs to FlowRateField + modal

IanLondon committed:
add id to  TerminalItem and StepCreationButton

Commit:
Merge remote-tracking branch 'origin/pd_add-e2e-selectors-7384' into pd_batch-edit-e2e

IanLondon committed:
test(cypress): update Cypress to 6.6.0

Commit:
Merge remote-tracking branch 'origin/js_update-cypress-6.6.0' into pd_batch-edit-e2e

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
initial e2e

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
adding the mixed step verification

Commit:
adding the comments

Commit:
formatting json batch edit

Commit:
moving json to 5 folder

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,78 @@
+    openDesignTab()
+
+    // enter into the batch edit mode
+    cy.get('[data-test="StepItem_1"]').click({
+      ctrlKey: true,
+    })
```
after this can we guarantee that the banner shows up or something? just to show we are in batch edit mode

reachojhaneha replied:
Good Point. I am doing set of actions in this script but not verifying them. Will verify these steps.

shlokamin replied:
cool, lmk if you'd like any support. and thanks for this, we really appreciate it!

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,78 @@
+    })
+
+    // Range selection with shift
+    cy.get('[data-test="StepItem_3"]').click({
+      shiftKey: true,
+    })
```
can we validate that the steps actually get selected? maybe via styles or something, we can look into this

reachojhaneha replied:
will do

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,78 @@
+    })
+
+    // Change the Flowrate to 100 and Save
+    cy.get('[name="aspirate_flowRate"]').click()
+    cy.get('input[name="aspirate_flowRate_customFlowRate"]').type('100')
+    cy.get('button').contains('Done').click()
+    cy.get('button').contains('save').click()
+
```
can we validate that he flow rate gets saved? maybe by reopening the form and making sure the value has changed

reachojhaneha replied:
will do

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,78 @@
+      }
+    )
+
+    // Dublicate the selected steps
+    cy.get('#ClickableIcon_duplicate').click()
+
+    // Expand ALL steps
+    cy.get('#ClickableIcon_expand').click()
+
+    // Delete the dublicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
+
+    // Exit batch edit mode
+    cy.get('button').contains('exit batch edit').click()
```
Same thing with all of these, we should validate that the steps have been deleted/duplicated/expanded. And that we've successfully exited batch edit mode

reachojhaneha replied:
will do

shlokamin reviewed (changes requested):
Prior to release we should make sure we assert that each of the actions we're doing were successful. I'm down to pitch in here!

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

IanLondon committed:
refactor(protocol-designer): add selectors for e2e tests

- Add SubstepRow_aspirateWell, SubstepRow_volume, and SubstepRow_dispenseWell data-test attrs
- Add StepSelectionBannerComponent_numberStepsSelected id

Commit:
Merge branch 'pd_steplist-e2e-selectors' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
adding assertions

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
spacing in commented code

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,112 @@
+    })
+    cy.get('input[name="aspirate_flowRate"]').should('have.value', 100)
+
+    // Range selection with shift
+    cy.get('[data-test="StepItem_3"]').click({
+      shiftKey: true,
+    })
```
how come this range selection is needed? it looks like just after there is another range selection on `StepItem_4`

reachojhaneha replied:
removed this step

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,112 @@
+      .contains('9 steps selected')
+      .should('exist')
+
+    // Dublicate the selected steps
+    cy.get('#ClickableIcon_duplicate').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('9 steps selected')
+      .should('exist')
+
+    // Delete the dublicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
```
this doesn't quite validate that a duplicate and delete went through. could we also verify that when you click `duplicate` there will be a total of 18 steps? could just be: 

```js
cy.get('[data-test="StepItem_18"]').should('exist') //there are now 18 step items
```

and then after the `delete`:

```js
cy.get('[data-test="StepItem_10"]').should.not.('exist') // there are now only 9 stepitems
```

or something like that... 

does that make sense?

reachojhaneha replied:
this is addressed, not after duplicating i verify that 18 steps are there and deleting steps verify the remaining steps too.

Commit:
verifying delete and dublicate step

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -71,13 +66,21 @@ describe('Batch Edit Transform', () => {
+      shiftKey: true,
+    })
+
     // Delete the dublicated steps
```
`dublicated  => duplicated` 😄

reachojhaneha replied:
Test is not finding exit batch edit mode button, as its not able to enter the batch edit form. Does the E2E run on production environment? As this case passes on sandbox

shlokamin replied:
no they only run on sandbox

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,115 @@
+  })
+
+  // import the batchEdit.json to PD
+  it('Verify Flowrate, dublicate, delete functionality in Batch Edit Mode', () => {
```
`dublicate  => duplicate` 😄

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,115 @@
+      .contains('9 steps selected')
+      .should('exist')
+
+    // Dublicate the selected steps
```
`dublicate => duplicate 😄`

reachojhaneha replied:
Thanks, will correct this

shlokamin reviewed (commented):
Some small typos, but it also looks like CI is failing. Looks this test cannot find the `exit batch edit mode` button

Commit:
os based key change

shlokamin commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,117 @@
+    // Duplicate the selected steps
+    cy.get('#ClickableIcon_duplicate').click()
+    cy.get('#ClickableIcon_select').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('18 steps selected')
+      .should('exist')
+
+    cy.get('[data-test="StepItem_9"]').click({
+      shiftKey: true,
+    })
+
+    // Delete the duplicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('1 steps selected')
+      .should('exist')
```
This part I do not follow...

1. We duplicated our 9 steps, so now we have 18 steps selected
2. We shift + click on step number 9 (at this point we would have deselected 9 steps, so 9 should now be selected right?)
3. Delete the selected steps (which theoretically is 9 steps)
4. Then only 1 item is selected

How is it that only 1 item is left selected? Is this because after we delete, we are always only left with one selected item?

shlokamin replied:
i think i answered my own q above...

would it be worth checking that we only have 9 steps left after clicking `delete`?

IanLondon replied:
Might be nice to have a comment just pointing out that after the shift-click, Steps 1 to 9 should be selected (bc it's hard to follow in my head as a read it)

Also agree I think we might want to delete, then select all, and confirm that 9 steps are remaining (instead of, or in addition to, checking that 1 step is currently selected immediately after clicking delete)

reachojhaneha replied:
this is what i tried: selected all the steps and shift + step 9 to deselect steps 9 to 18. Delete steps 1-8 steps as they are selected. Once all the selected steps are deleted, next step becomes selected as for batch edit one step is always selected(imo this is a verification too).
Now, about selecting steps 9-18 steps, select-all(id - #ClickableIcon_select)  is hidden by another element <div class="lists__title_bar__3n-Et lists__clickable__2o_y_">...</div>, and cypress is not able to access it. Hence i took the other route mentioned above.

IanLondon commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,117 @@
+      }
+    )
+
+    // Expand ALL steps
```
I think this comment is inaccurate, it's really expanding just the selected steps (1 and 4 at this point)

IanLondon commented on protocol-designer/cypress/integration/batchEdit.spec.js:
```diff
@@ -0,0 +1,117 @@
+
+    // Expand ALL steps
+    cy.get('#ClickableIcon_expand').click()
+    cy.get('[data-test="SubstepRow_aspirateWell"]')
```
This is testing that the step is expanded, right? Maybe we should make sure specifically that steps 1 & 4 are expanded and that eg step 2 isn't? Right now, this would pass if any one step was expanded, I think.

IanLondon reviewed (changes requested):
Looks great! Just have a few comments.

Also, could you merge the latest edge on to this? It has some changes that have already been merged into edge (eg StepSelectionBanner + SubstepRow) and also since this branch may be getting a little older, it'd be good to update it so that we're not surprised by something unexpected happening when it is merged

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
changing ctrlKey

Commit:
verification

shlokamin commented:
Looks like e2e cases are failing: 

```
AssertionError: Timed out retrying after 4000ms: Expected to find content: 'Batch editing of settings is only available for Transfers' within the element: <p#Text_noSharedSettings.Text-sc-1wb1h0f-0.nVDCX> but never did.
```

Kadee80 commented:
> Looks like e2e cases are failing:
> 
> ```
> AssertionError: Timed out retrying after 4000ms: Expected to find content: 'Batch editing of settings is only available for Transfers' within the element: <p#Text_noSharedSettings.Text-sc-1wb1h0f-0.nVDCX> but never did.
> ```

Might this be because we changed the text to Transfers or Mix?

shlokamin commented:
Ahhh yep, we have to change the cypress query to `"Batch editing of settings is only available for Transfer or Mix steps"`

Commit:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pd_batch-edit-e2e

Commit:
copy change

IanLondon reviewed (approved):
🎉
